### DISCUSSION TITLE
hyprland load bind config before hyprpm load plugin?

### DISCUSSION THREAD

**Discussion by DreamMaoMao** [hyprwm/Hyprland#4373] [General]
![image](https://github.com/hyprwm/Hyprland/assets/30348075/5af214f0-a7d7-489e-8936-24eeda9266a5)
I need some help.

![image](https://github.com/hyprwm/Hyprland/assets/30348075/3b608368-f6f2-4d6f-b44a-c69af2c2f723)

Some of the custom dispatch commands on my plugin load before hyprpm reload, so it starts with an error.

how to fix it??

**Comment by vaxerski** [comment]
well hyprpm reload gets launched as a command, and it doesn't reload the config, your plugins should do that, in the init func.

A workaround would be to change the exec once to:
```ini
exec-once = hyprpm reload && hyprctl reload
```

**Comment by DreamMaoMao** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/30348075/90a98825-e25d-4183-b6aa-1393f2643504)

I have done this, but the boot always appears an error message, and then the error message disappears, this boot way is too ugly

**Comment by DreamMaoMao** [comment]
I want error messages to never appear, which requires the bind configuration to load after hyprpm loads the plug-in

**Comment by DreamMaoMao** [comment]
I hope you consider this need, this kind of startup is really not very beautiful
@vaxerski

**Reply by vaxerski** [reply]
what if you add the exec-once thing I showed? Worth noting you can also disable error messages in the config.

**Comment by DreamMaoMao** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/30348075/5b97d220-e399-4943-9560-8a11d06145cf)

Same phenomenon, the error message appears first, and then the error message disappears

**Reply by vaxerski** [reply]
I mean, it's hard to really make it be gone, as `hyprpm reload` is executed as a subprocess, and then hyprland carries on.

What I'd recommend I guess is just disabling the error messages for the first few seconds and then re-enabling it

**Reply by DreamMaoMao** [reply]
Well, there doesn't seem to be a better way

**Reply by dskvr** [reply]
@vaxerski What about [this](https://github.com/hyprwm/Hyprland/discussions/4373#discussioncomment-14677846)?

**Comment by DreamMaoMao** [comment]
> what if you add the exec-once thing I showed? Worth noting you can also disable error messages in the config.

I think it's a little rude to just disable the error messag

**Comment by ArneshRC** [comment]
~~```exec-once = hyprpm reload -n && hyprctl keyword source ~/.config/hypr/binds.conf```~~

~~This works for me.~~

Edit: Nevermind, it doesn't.

**Comment by dskvr** [comment]
I know this topic is closed, but from both a UX perspective and a maintainer's perspective the suggested workarounds are subpar. I couldn't find any evidence of proposed solutions to this issue, either.

I have thought about this quite a bit and took a [stab at solving this race condition](https://github.com/sandwichfarm/Hyprland/pull/1). IMO it is a very elegant solution in a minimal changeset. **It is not a complete solution, rather a proof of concept, but I believe is a start to a complete solution.** _However, as was true previously with other topics, I may overlooked something._ 

IMO The reason this doesn't sit right with me (and others) is error fatigue and the externalization of internal logic on to users, as well as the need for plugin developers to communicate these race-conditions to a group a users with varying skill and knowledge levels; which adds friction to development.

<details>
<summary>
Truncated PR Description
</summary> 

## Requirements: 
1. Low maintenance
2. No breaking changes
3. No regressions to existing configurations or plugins.
4. 30 minute timebox

## UX and development issues resolved:
1. `errors == errors == true.`; Users are only shown actual errors, not side-effects of internal race conditions
2. Eliminate the externalization of internal logic on to users; **user's shouldn't need to remember if they have errors turned on or off.**

**I have successfully tested and validated the following cases:**
1. Valid dispatcher for valid plugin; successfully defers and the error is dismissed without displaying
2. Invalid dispatcher for valid plugin; successfully defers and is thrown once plugin is loaded
3. Dispatcher for invalid (non-existent) plugin; successfully defers and is thrown once config is deemed invalid. 
</details>

Instructions to run the tests are available in the PR. _Will open a PR to upstream if requested._

**Reply by aaronjamt** [reply]
I'll add my +1 on this, I'd love to stop seeing errors on boot while plugins load. I doubt vaxry will see this here, so personally I'd just open the PR and let him close it if he doesn't agree.

**Reply by dskvr** [reply]
I've already bothered him enough. If this gets enough upvotes eventually someone will share it with him.
